### PR TITLE
Slerp + dict of interpolation functions

### PR DESCRIPTION
**Pull request by Fifourche**
This PR is to propose a way for the user to give custom interpolation functions to compute frames. It also contains a Slerp default method. A related issue has been opened at #37 .

## About custom interpolation functions
In details, I added a "methods" argument to the `capture_keyframe` function. "methods" is a dict of dicts containing functions, normally of same structure than "state". 

For instance, the custom interpolation function for the `state['camera']['angles']` would be looked for at `methods['camera']['angles']`.

A few things to notice and to be discussed I think :

- I made it so if an user wants to provided only a custom function for the "zoom" attribute for instance, he wouldnt' have to fill a whole dict of dicts ; he could just pass `methods = {'camera': {'zoom': custom_function}}`. Other attributes' interpolation functions would be defaulted.
- an interpolation function is expected to be of the form `custom_func(a, b, fraction)`
- it is possible to pass a function for a whole subpart of "state" ; i.e. if the interpolation should intertwine all "camera" attributes, we can pass `methods = {'camera': custom_func}`. 
`custom_func` will take `initial_state['camera']` and `final_state['camera']` dicts as `a` and `b` arguments

## About Slerp
I took advantage of this to add a Slerp method, relying on `scipy`. The default interpolation function for "angles" is set to Slerp.

### An example code :
Here's an example code where "methods" is either set to a standard linear interpolation for "angles" and or defaulted, resulting in Slerp for "angles".
Playing with the slider will help visualizing the difference.

```Python
import napari
from napari_animation import AnimationWidget
from skimage.data import cells3d

cells = cells3d()[:, 1]  # get some data

# Launch viewer to view data
viewer = napari.view_image(cells, colormap='magma')
viewer.dims.ndisplay = 3

# Instantiate the animation widget
animation_widget = AnimationWidget(viewer)
viewer.window.add_dock_widget(animation_widget, area='right')
animation = animation_widget.animation

# Angles interpolation is set to simple linear interpolation of a sequence
from napari_animation.utils import _interpolate_seq
methods_test = {'camera': {'angles': _interpolate_seq}}
# methods_test = {} # uncomment to default to Slerp instead

#%% Capture frames

animation.key_frames.clear()

viewer.camera.angles = (0, 0, 90)
animation.capture_keyframe(methods=methods_test)

viewer.camera.angles = (0, 0, 180)
animation.capture_keyframe(methods=methods_test)

viewer.camera.angles = (-180, -90, 0)
animation.capture_keyframe(methods=methods_test)

viewer.camera.angles = (180, 0, 0)
animation.capture_keyframe(methods=methods_test)

viewer.camera.angles = (-180, 90, 0)
animation.capture_keyframe(methods=methods_test)

viewer.camera.angles = (0, 0, -180)
animation.capture_keyframe(methods=methods_test)

viewer.camera.angles = (0.0, 0.0, 90.0)
animation.capture_keyframe(methods=methods_test)
```

### CODE REVIEW

**Review by alisterburt — COMMENTED** [review]
Awesome PR @Fifourche - thanks so much for the time and effort!!

There are some things I like about this and some I like a bit less, I'd be really interested to hear your thoughts and what @sofroniewn and @guiwitz think about the following...

- I like providing a custom interpolation function with the signature `f(a, b, fraction)` as you suggested more than providing separate 'forward' and 'inverse' transforms prior to standard interpolation as suggested by @sofroniewn in #37. I think for more complicated transforms (like slerp and the relevant euler2quaternion stuff in this PR) it's easier to reason about and keep tidy... happy to be told I'm wrong here though!

- I'm not so sure on this being exposed as a keyword argument in the `Animation.capture_keyframe()`, I think it should be managed managed by us and invisible to the user. I can't imagine a situation where I want to manually set these to something different than the 'correct' function for a given attribute

- the cascading functionality is nice but I can't imagine a situation where I want one custom function for multiple attributes, did you have a specific use case in mind?

- I think I prefer the notation @sofroniewn used to define attributes `'camera.zoom'` etc
```Python
{
'camera.zoom':{'forward': lambda x: np.power(10, x), 'inverse':  np.log10)},
'camera.angles':{'forward': some_function, 'inverse': some_inverse_function},
}
```

---

Given what I've said above I think what we might want to do is
- create an `Interpolation` enum like we do for easing functions, in a new module called `interpolation.py`, this should contain all current and future interpolation functions, moving them out of `utils.py`

Easing func enum as an example is here
 https://github.com/napari/napari-animation/blob/829bc31cbd222453b08157d7a77d544974e21fc9/napari_animation/easing.py#L242-L289

- create a dict of form
```python
{
'camera.zoom' : Interpolation.LOG,
'camera.angles' : Interpolation.SLERP
}
```
for attributes which require a custom interpolation function

- add a mechanism for parsing that dict and using the specific functions on the specified attribute

We could also drop the enum and simple have a dict of functions - curious to hear your thoughts!!

**Comment by sofroniewn** [comment]
I'm afraid I havn't got too much time right now to think deeply about this, but I like the idea of getting rid of my `pow 10` and `log` functions for zoom etc. that really did feel like a hack. I also tend to agree with a lot of @alisterburt points about API, keeping this hidden from users as much as possible if we know what should be done for an attribute, and using an API something like 
```python
{
'camera.zoom' : Interpolation.LOG,
'camera.angles' : Interpolation.SLERP
}
```
which seems quite elegant.

I think this PR might also contain some changes from #61, so once we get that merged this should be easier to review. I should be able to get to it early next week, but in the mean time I say press on! 🚀

**Comment by Fifourche** [comment]
Thanks for your thoughts ! :) I agree with you, it should probably stay invisible to the user !

For this and for the cascading, what I had in mind at the time was the interpolation of the recently added layers' states. Some of the properties -not the base ones, but things like colormap or others- might be tricky to interpolate, and I imagined scenarios where the interpolation fucntion would need to change opacity as well as other parameters for instance.

It was and still isn't clear in my mind how to prepare for this, but in any case I agree with everything you said above, and will adopt the proposed form ! Thanks again 😃

**Comment by alisterburt** [comment]
awesome - look forward to seeing what you come up with @Fifourche !

You should join us on the [napari zulip chat](https://napari.zulipchat.com/) - there are also developer meetings each week, it would be great to have you there!

**Comment by Fifourche** [comment]
I made some changed according to the discussion !

I created a dictionary :
```Python
interpolation_dict = {
    "camera.angles": Interpolation.SLERP,
    "camera.zoom": Interpolation.LOG,
}
```
The unspecified properties will be defaulted to `Interpolation.LINEAR`. One thing I don't know is where to place this dict ; I put it at the very end of `interpolation.py` but it may be a bit hidden there.

Also, I guess some tests should be changed accordingly ; I'll take care of that soon if you're ok with the commits so far !

File: napari_animation/interpolation.py
Comment (sofroniewn) [inline napari_animation/interpolation.py:12]: Can you add `Parameters` and `Returns` sections to these doc strings

File: napari_animation/interpolation.py
Comment (sofroniewn) [inline napari_animation/interpolation.py:9]: I guess this could be ok for now, but I'd prefer if we didn't need the direct vispy import here. Maybe there is quaternion support inside scipy, or we can just vendor/ define the functions/ objects we need in a utils file here.

The reason is one day napari might support other rendering engines and so I'd like to keep our coupling to vispy as light as possible. This seems like one place where we could have a vispy independent alternative

File: napari_animation/interpolation.py
Comment (sofroniewn) [inline napari_animation/interpolation.py:91]: I wonder if we shouldn't move these to a different file/ locations, maybe put them directly onto our animation object. They could either be private, or I could imagine someone might want to overwrite them, add their own custom attributes to this dictionary and having them on `Animation` seems like a good place.

File: napari_animation/utils.py
Comment (sofroniewn) [inline napari_animation/utils.py:50]: Can you add `Parameters` and `Returns` sections to the doc strings here and to the other nested functions

File: napari_animation/utils.py
Comment (sofroniewn) [inline napari_animation/utils.py:5]: This might make sense to put inside the `interpolation.py` file now, but I don't really mind either way

**Review by sofroniewn — APPROVED** [review]
Looking really great! I've left mostly minor, cosmetic comments, flagged a couple doc strings that need a few more sections, but otherwise this looks great and I think will then be good to merge @alisterburt when you're ready.

I'd also like to second @alisterburt comment about the [napari zulip](https://napari.zulipchat.com/login/) and communities meetings. I see you're in the zulip now, so I'll reach other there and give you additional details on the meetings.

File: napari_animation/interpolation.py
Comment (Fifourche) [reply napari_animation/interpolation.py:12]: I added some to all these functions, it might be silly but just in case. ^^

File: napari_animation/interpolation.py
Comment (Fifourche) [reply napari_animation/interpolation.py]: For now, I left it in `interpolation.py` ! I agree that we might want to define it as an `Animation` attribute and pass it as a parameter to `interpolate_state`, if there's no objection !

Comment (sofroniewn) [reply napari_animation/interpolation.py]: @alisterburt what do you think? Agreed it's something that can come later, so we can merge this first and see

File: napari_animation/interpolation.py
Comment (sofroniewn) [reply napari_animation/interpolation.py:12]: No it's really great! They should all have doc strings. We'll start building docs and sharing them online soon and then it will be really nice to have complete coverage in the standard form. Thanks!!

**Comment by sofroniewn** [comment]
@Fifourche we've got some merge conflicts now since I merged #68, do you mind updating this PR. Then I think it is good to go!!

File: napari_animation/interpolation.py
Comment (Fifourche) [reply napari_animation/interpolation.py]: I'm ok with `state_interpolation_map`, and added it ! Should be ok now :)

Comment (alisterburt) [reply napari_animation/interpolation.py]: you're on fire! Will have a quick play then get to merging

**Review by alisterburt — APPROVED** [review]
Merge time 🚀

**Pull request merged**
